src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,433 @@
+
+        /// <summary>
+        ///   The local peer's end point.
+        /// </summary>
+        public MultiAddress LocalAddress { get; set; }
+
+        /// <summary>
+        ///   The remote peer's end point.
+        /// </summary>
+        public MultiAddress RemoteAddress { get; set; }
+
+        /// <summary>
+        ///   The private key of the local peer.
+        /// </summary>
+        /// <value>
+        ///   Used to prove the identity of the <see cref="LocalPeer"/>.
+        /// </value>
+        public Key LocalPeerKey { get; set; }
+
+        /// <summary>
+        ///   Determine which peer (local or remote) initiated the connection.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> if the <see cref="RemotePeer"/> initiated the connection;
+        ///   otherwise, <b>false</b>.
+        /// </value>
+        public bool IsIncoming { get; set; }
+
+        /// <summary>
+        ///   Determines if the connection to the remote can be used.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> if the connection is active.
+        /// </value>
+        public bool IsActive => Stream != null && Stream.CanRead && Stream.CanWrite;
+
+        /// <summary>
+        ///   The duplex stream between the two peers.
+        /// </summary>
+        public Stream Stream
+        {
+            get => _stream;
+            set
+            {
+                if (value != null && _statsStream == null)
+                {
+                    _statsStream = new StatsStream(value);
+                    value = _statsStream;
+                }
+
+                _stream = value;
+            }
+        }
+
+        /// <summary>
+        ///   The protocols that the connection will handle.
+        /// </summary>
+        /// <value>
+        ///   The key is a protocol name, such as "/mplex/6.7.0".  The value
+        ///   is a function that will process the protocol message.
+        /// </value>
+        /// <seealso cref="AddProtocol"/>
+        /// <seealso cref="AddProtocols"/>
+        public Dictionary<string, Func<PeerConnection, Stream, CancellationToken, Task>> Protocols { get; }
+            = new Dictionary<string, Func<PeerConnection, Stream, CancellationToken, Task>>();
+
+        /// <summary>
+        ///   Add a protocol that the connection will handle.
+        /// </summary>
+        /// <param name="protocol">
+        ///   A peer protocol to add.
+        /// </param>
+        public void AddProtocol(IPeerProtocol protocol)
+        {
+            Protocols.Add(protocol.ToString(), protocol.ProcessMessageAsync);
+        }
+
+        /// <summary>
+        ///   Add a seequence of protocols that the connection will handle.
+        /// </summary>
+        /// <param name="protocols">
+        ///   The peer protocols to add.
+        /// </param>
+        public void AddProtocols(IEnumerable<IPeerProtocol> protocols)
+        {
+            foreach (var protocol in protocols)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)